cpp/src/arrow/util/decimal.cc

```diff
@@ -253,117 +251,131 @@ static void StringToInteger(const std::string& str, Decimal128* out) {
   }
 }
 
-static const boost::regex DECIMAL_REGEX(
-    // sign of the number
-    "(?<SIGN>[-+]?)"
-
-    // digits around the decimal point
-    "(((?<LEFT_DIGITS>\\d+)\\.(?<FIRST_RIGHT_DIGITS>\\d*)|\\.(?<SECOND_RIGHT_DIGITS>\\d+)"
-    ")"
+namespace {
 
-    // optional exponent
-    "([eE](?<FIRST_EXP_VALUE>[-+]?\\d+))?"
+struct DecimalComponents {
+  std::string sign;
+  std::string whole_digits;
+  std::string fractional_digits;
+  std::string exponent_sign;
+  std::string exponent_digits;
+};
 
-    // otherwise
-    "|"
+inline bool IsSign(char c) { return (c == '-' || c == '+'); }
 
-    // we're just an integer
-    "(?<INTEGER>\\d+)"
+inline bool IsDot(char c) { return c == '.'; }
 
-    // or an integer with an exponent
-    "(?:[eE](?<SECOND_EXP_VALUE>[-+]?\\d+))?)");
+inline bool IsDigit(char c) { return (c >= '0' && c <= '9'); }
 
-static inline bool is_zero_character(char c) { return c == '0'; }
+inline bool StartsExponent(char c) { return (c == 'e' || c == 'E'); }
 
-Status Decimal128::FromString(const std::string& s, Decimal128* out, int32_t* precision,
-                              int32_t* scale) {
-  if (s.empty()) {
-    return Status::Invalid("Empty string cannot be converted to decimal");
+inline size_t ParseDigitsRun(const char* s, size_t start, size_t size, std::string* out) {
+  size_t pos;
+  for (pos = start; pos < size; ++pos) {
+    if (!IsDigit(s[pos])) {
+      break;
+    }
   }
+  *out = std::string(s + start, pos - start);
+  return pos;
+}
 
-  // case of all zeros
-  if (std::all_of(s.cbegin(), s.cend(), is_zero_character)) {
-    if (precision != nullptr) {
-      *precision = 0;
-    }
+bool ParseDecimalComponents(const char* s, size_t size, DecimalComponents* out) {
```

Comment:
Why not use iterators instead of a pointer plus size? It seems like we're subverting the language.

Reply:
I find it much readable like that. I don't know what "subverting the language" means.

Reply:
Oh, I just mean that iterators abstract over pointer arithmetic and are built into the STL. It feels like C as opposed to C++, that's all.

Reply:
I'll try an iterator version tomorrow and see if I like it better. It's a bit late tonight ;-)

Reply:
By no means should you feel obliged to do that extra work :) If this works, then that's great. I thought maybe you had chosen not to use iterators for a particular reason.

Reply:
I've decided against this. Not only would it be slightly less readable and more annoying to type, but the day we have a string view or span, the current solution adapts seamlessly. The "abstracted" variant OTOH is less abstracted from that point of view.

Reply:
I don't follow how a `view`/`span` class is less abstracted, since that would presumably implement the c++ iterator interface, like every implementation of it usually does. However, as I said, I don't think this is really worth spending too much time on.

Reply:
I mean if the parse function is taking a `string::const_iterator`, it maybe won't accept a different kind of iterator. Or we need to make it a template function, piling more layers of abstraction without any concrete advantage.